engine/env/spark/src/main/java/org/datacleaner/spark/functions/ExtractJsonInputRows.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * DataCleaner (community edition)
+ * Copyright (C) 2014 Neopost - Customer Information Management
+ *
+ * This copyrighted material is made available to anyone wishing to use, modify,
+ * copy, or redistribute it subject to the terms and conditions of the GNU
+ * Lesser General Public License, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this distribution; if not, write to:
+ * Free Software Foundation, Inc.
+ * 51 Franklin Street, Fifth Floor
+ * Boston, MA  02110-1301  USA
+ */
+package org.datacleaner.spark.functions;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.metamodel.data.DataSet;
+import org.apache.metamodel.data.Row;
+import org.apache.metamodel.json.JsonDataContext;
+import org.apache.metamodel.query.Query;
+import org.apache.metamodel.schema.Column;
+import org.apache.metamodel.schema.Table;
+import org.datacleaner.connection.JsonDatastore;
+
+/**
+ * Reads and extracts data from a Json datastore
+ *
+ */
+public class ExtractJsonInputRows {
+
+    final JsonDatastore _jsonDatastore;
+
+    public ExtractJsonInputRows(JsonDatastore jsonDatastore) {
+        _jsonDatastore = jsonDatastore;
+    }
+
+    public List<Object[]> getInputRows() {
+        final List<Object[]> list = new ArrayList<>();
+        final JsonDataContext jsonContext = new JsonDataContext(_jsonDatastore.getResource());
+        final Table table = jsonContext.getDefaultSchema().getTable(0);
+        final Column[] columns = table.getColumns();
+        final Query query = new Query().select(columns).from(table);
+        final DataSet executeQuery = jsonContext.executeQuery(query);
```

Comment:
This will make MetaModel read the file on HDFS from end to end. That basically means we don't get any scaling behaviour because the read is not distributed. Worse, it also means that every node will do the exact same thing, creating duplicated results.

Let's talk about this on Skype, but in short you have to use the SparkContext to read the text file line-by-line and pass each line further into the processing steps. If you don't do that you also build up the complete file in memory which isn't feasible either.